core/multi_map.go

```
@@ -0,0 +1,143 @@
+type MultiMap interface {
+	// IDistributedObject is the base interface for all distributed objects.
+	IDistributedObject
+
+	// Stores a key-value pair in the multi-map.
+	// It returns true if size of the multi-map is increased, false if the multi-map
+	// already contains the key-value pair.
+	Put(key interface{}, value interface{}) (increased bool, err error)
+
+	// Get returns a slice of values associated with the specified key.
+	Get(key interface{}) (values []interface{}, err error)
+
+	// Remove removes an association of the specified value with the specified key. Calling this method does not affect
+	// other values associated with the same key.
+	// It returns true if the value was detached from the specified key, false if it was not.
+	Remove(key interface{}, value interface{}) (removed bool, err error)
+
+	// RemoveAll detaches all values from the specified key.
+	// It returns a slice of old values that were associated with this key prior to this method call.
+	RemoveAll(key interface{}) (oldValues []interface{})
+
+	// ContainsKey checks if this multi-map contains a specified key.
+	// It returns true if this map contains the specified key, false otherwise.
+	ContainsKey(key interface{}) (found bool, err error)
+
+	// ContainsValue returns true if the specified value is associated with at least one key in this multi-map,
+	// false otherwise.
+	ContainsValue(value interface{}) (found bool, err error)
+
+	// ContainsEntry returns true if this multi-map has an association between
+	// the specified key and the specified value, false otherwise.
+	ContainsEntry(key interface{}, value interface{}) (found bool, err error)
+
+	// Clear removes all entries from this multi-map.
+	Clear() (err error)
+
+	// Size returns the total number of values in this multi-map.
+	Size() (size int32, err error)
+
+	// ValueCount returns the number of values associated with the specified key.
+	ValueCount(key interface{}) (valueCount int32)
```

This method should also return an error